File: bridge/python/tests/parse.py
from session import open_workspace


def test_parse_returns_clean_module() -> None:
    fixture = open_workspace(
        [
            ("destack.json", '{"name":"@test/app"}'),
            ("src/index.ds", "export const value = 1;"),
        ],
    )
    workspace = fixture.workspace
    revision = workspace.revision()
    module = workspace.module("src/index.ds")

    # parse one loaded source module
    output = workspace.parse(revision, module)

    assert output.diagnostics == []


def test_parse_returns_syntax_error() -> None:
    fixture = open_workspace(
        [
            ("destack.json", '{"name":"@test/app"}'),
            ("src/index.ds", "export const broken ="),
        ],
    )
    workspace = fixture.workspace
    revision = workspace.revision()
    module = workspace.module("src/index.ds")

    # parse and inspect the returned syntax diagnostic
    output = workspace.parse(revision, module)

    assert len(output.diagnostics) == 1
    diagnostic = output.diagnostics[0]
    assert diagnostic.code == "EP001"
    assert diagnostic.severity.label == "error"
    assert diagnostic.message == "parse error: unexpected End in Declarator"

File: bridge/rust/tests/parse.rs
#[path = "support/session.rs"]
mod session;

use session::open_workspace;

#[test]
fn test_parse_returns_clean_module() -> destack::Result<()> {
    let fixture = open_workspace(&[
        ("destack.json", r#"{"name":"@test/app"}"#),
        ("src/index.ds", "export const value = 1;"),
    ])?;
    let workspace = fixture.workspace;
    let revision = workspace.revision()?;
    let module = workspace.module("src/index.ds")?;

    // parse one loaded source module
    let output = workspace.parse(revision, module)?;

    assert_eq!(output.diagnostics, []);

    Ok(())
}

#[test]
fn test_parse_returns_syntax_error() -> destack::Result<()> {
    let fixture = open_workspace(&[
        ("destack.json", r#"{"name":"@test/app"}"#),
        ("src/index.ds", "export const broken ="),
    ])?;
    let workspace = fixture.workspace;
    let revision = workspace.revision()?;
    let module = workspace.module("src/index.ds")?;

    // parse and inspect the returned syntax diagnostic
    let output = workspace.parse(revision, module)?;
    let diagnostic = &output.diagnostics[0];

    assert_eq!(output.diagnostics.len(), 1);
    assert_eq!(diagnostic.code, "EP001");
    assert_eq!(diagnostic.severity, destack::DiagnosticSeverity::Error);
    assert_eq!(
        diagnostic.message,
        "parse error: unexpected End in Declarator"
    );

    Ok(())
}
